### ISSUE TITLE
RFE: Automatically determine smartcard reader name

### ISSUE THREAD

**Issue by opoplawski** [rdesktop/rdesktop#266]
It seems we can determine the available reader from SCardListReaders so we shouldn't have to force the user to pass -o sc-reader-name="...".

**Comment by uglym8** [comment]
Actually, atm you don't have to specify reader name at all.
You may, but you don't have to.

Try to use only **"-r scard"**.

Later when the #239 will be fixed, this (i.e. passing all readers or just particular reader) may become mandatory argument to **-r scard** option.

Regarding getting reader name from SCardListReaders:

yes, we can get the name (names probably), but there are some caveats:

1) sometimes there are more than one reader connected (e.g. USB token and external reader AND smartcard reader inside the keyboard). 
So you need a way to pass only required reader.
2) sometimes you want to use alias in windows for the smartcard reader.

**Comment by opoplawski** [comment]
If I only use "-r scard" I get:
```
Core(warning): iso_connect(), missing smartcard information for SSO, disabling CredSSP
Failed to connect, CredSSP required by server.
```
I understand the possible need to specify the reader, but in our case we only have the one.

**Comment by uglym8** [comment]
I guess you launching rdesktop with **"-i"** option?

I overlooked this part (you actually said about passing **-o sc-reader-name="..."**).
My apologies.

So, yes in this case (supplying **"-o sc-reader-name="**) we definitely can get the name from SCardListReaders() if the user doesn't supply any of the **sc-*** variables.

In case when you don't use smartcard as authentication token (i.e. not supplying **"-i"** option) you can just pass "**-r scard"** to redirect all smartcard readers.

**Comment by opoplawski** [comment]
Ah, yes, I am passing in "-i" as well.  Sorry I didn't mention that.

Also, on initial connection I get "The requested key container does not exist on the smart card".  Any idea how to fix that?  I can dismiss it and choose the right user/pin, but it would be nice to authenticate directly.

**Comment by hean01-cendio** [comment]
@opoplawski You need to provide a crypto service provider (CSP) string using -o sc-csp-name="CSP Provider Name" argument to rdesktop. The CSP is part of your smart card driver,  you will find the names for available CSP's in the registry 'HKLM\SOFTWARE\Microsoft\Cryptography\Defaults\Provider' on the RDS server. Identify the one matching provider string for your smart card driver and use in argument to rdesktop.

**Comment by uglym8** [comment]
Have you ever been able to redirect this card before?
Can you see the reader?
What software do you use to check whether the card works?

**Comment by opoplawski** [comment]
Another data point - redirection is still working with a Windows Server 2012R2 host, but not with a Windows 10 Ent host.  Before I used sc-csp-name, while direct authentication failed, the machine could at times see the card and authenticate that way.  Failure indications: certutil -sciinfo would fail and screen lock would not be able to use the smart card to unlock.

**Comment by opoplawski** [comment]
And now scard redirection is working just fine.  Something in the chain must be a little flaky.

**Comment by opoplawski** [comment]
When it does go south, we see this on the remote end:
```
An error occurred while retrieving a digital certificate from the inserted smart card. The Smart Card Resource Manager is not running.
```
often after disconnecting and reconnecting a session.  Probably should start a new issue for this though...

**Comment by uglym8** [comment]
So, to sum up:
1) It seems that there's no point in getting reader name via SCardListReaders() to fill in the **sc-reader-name** as it has turned out that sometimes (@hean01-cendio:or always?) you have to specify **sc-csp-name** to authenticate using smartcard.
2) There're some issues (of origin yet to be found) with smartcard authentication for which new issue will be created.

@opoplawski, @hean01-cendio: 
Sounds right, guys?

**Comment by opoplawski** [comment]
Yeah, setting sc-reader-name doesn't seem that necessary now.  Setting --sc-csp-name does though, and it seems like it might be good to default to "Microsoft Base Smart Card Crypto Provider".

**Comment by hean01-cendio** [comment]
>It seems that there's no point in getting reader name via SCardListReaders() to fill in the sc-reader-name as it has turned out that sometimes (@hean01-cendio:or always?) you have to specify sc-csp-name to authenticate using smartcard.

Yes this is correct.

>Yeah, setting sc-reader-name doesn't seem that necessary now. Setting --sc-csp-name does though, and it seems like it might be good to default to "Microsoft Base Smart Card Crypto Provider".

I agree to this, we should provide a default and that would be a good choice for csp. When another CSP is required an error will occur at logon and one need to provide the correct csp.

@uglym8, @opoplawski

**Comment by opoplawski** [comment]
So, what needs to get done next?
